### PR TITLE
Fix integer to string conversion when generating code

### PR DESCRIPTION
**Pull request by grendello**
Fixes: https://github.com/xamarin/xamarin-android/issues/7939
Context: https://github.com/dotnet/runtime/issues/13363
Context: https://www.unicode.org/charts/PDF/U2200.pdf (page 2)

LLVM IR generator outputs a number of integer values, relying on the
standard `ToString()` method to convert the value to string.  However,
since NET6, this conversion is culture-sensitive and in certain
cultures (list below) it will output the minus sign not as the standard
ASCII 0x2D character but rather as the Unicode 0x2212 (mathematical
operator "minus") character.  This breaks LLVM LLC:

    llc: environment.arm64-v8a.ll:554:7: error: expected value token
      stderr |                 i32 −1, ; apk_fd

Fix the problem by using invariant culture when converting **all**
integers and unknown objects to strings.  The reason all of them are
converted in this way is to avoid future changes to ICU and/or .NET
to-string conversion that might affect the resulting integer format.

Locales/cultures affected by this issue are as follows:

  * (RTL "-") 26 cultures
    * ar
    * ar-001
    * ar-BH
    * ar-DJ
    * ar-EG
    * ar-ER
    * ar-IL
    * ar-IQ
    * ar-JO
    * ar-KM
    * ar-KW
    * ar-LB
    * ar-MR
    * ar-OM
    * ar-PS
    * ar-QA
    * ar-SA
    * ar-SD
    * ar-SO
    * ar-SS
    * ar-SY
    * ar-TD
    * ar-YE
    * sd
    * sd-Arab
    * sd-Arab-PK
  * "‎-": 10 cultures
    * ar-AE
    * ar-DZ
    * ar-EH
    * ar-LY
    * ar-MA
    * ar-TN
    * he
    * he-IL
    * ur
    * ur-PK
  * (RTL "-"): 3 cultures
    * ckb
    * ckb-IQ
    * ckb-IR
  * "−": 38 cultures
    * et
    * et-EE
    * eu
    * eu-ES
    * fi
    * fi-FI
    * fo
    * fo-DK
    * fo-FO
    * gsw
    * gsw-CH
    * gsw-FR
    * gsw-LI
    * hr
    * hr-BA
    * hr-HR
    * ksh
    * ksh-DE
    * lt
    * lt-LT
    * nb
    * nb-NO
    * nb-SJ
    * nn
    * nn-NO
    * no
    * rm
    * rm-CH
    * se
    * se-FI
    * se-NO
    * se-SE
    * sl
    * sl-SI
    * sv
    * sv-AX
    * sv-FI
    * sv-SE
  * "‎−": 3 cultures
    * fa
    * fa-AF
    * fa-IR
  * "‎-‎": 16 cultures
    * ks
    * ks-Arab
    * ks-Arab-IN
    * lrc
    * lrc-IQ
    * lrc-IR
    * mzn
    * mzn-IR
    * pa-Arab
    * pa-Arab-PK
    * ps
    * ps-AF
    * ps-PK
    * ur-IN
    * uz-Arab
    * uz-Arab-AF

### CODE REVIEW

**Review by dellis1972 — APPROVED** [review]

**Review by jonathanpeppers — COMMENTED** [review]
Like I did in: https://github.com/dotnet/maui/commit/c40c6e773c48c995fcea28b71ee1dcf3a4bd1a1b

Can we add these rules as errors in the `.editorconfig`?
```
dotnet_diagnostic.CA1307.severity = error
dotnet_diagnostic.CA1309.severity = error
```

**Comment by grendello** [comment]
> Like I did in: [dotnet/maui@c40c6e7](https://github.com/dotnet/maui/commit/c40c6e773c48c995fcea28b71ee1dcf3a4bd1a1b)
> 
> Can we add these rules as errors in the `.editorconfig`?
> 
> ```
> dotnet_diagnostic.CA1307.severity = error
> dotnet_diagnostic.CA1309.severity = error
> ```

We should definitely do it, but in a separate PR, I think. It's not directly related to this PR.

**Comment by jonpryor** [comment]
@grendello: I would like a unit test of some form for this.  Could we alter one of the existing app build + deploy tests in `tests/MSBuildDeviceIntegration` to set the locale for the *entire build* process, including app deploy?

@dellis1972, @jonathanpeppers?

**Comment by jonathanpeppers** [comment]
For Mac, we could add test of a `Release` build, using the option to pass the `$LANG` env var:

https://github.com/xamarin/xamarin-android/blob/59e1e466bcb6739e976f61a65ce14a24bb6e3bf6/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/ProjectBuilder.cs#L63

But then we'd want to do some equivalent for Windows, too?

**Comment by jonathanpeppers** [comment]
Windows what I'm reading seems like a system-wide weird setting, someone recommends creating a user to run a process under...

So maybe it's just not worth it, and we make a new test that only runs on Mac.

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest2.cs [hunk 1]
```diff
@@ -105,14 +105,30 @@ public void CompressedWithoutLinker ()
...
-			using (var b = CreateApkBuilder ()) {
-				Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
+			string? oldLangVar = null;
+
+			if (!String.IsNullOrEmpty (langEnvironmentVariable)) {
+				// See: https://github.com/xamarin/xamarin-android/pull/7941
+				oldLangVar = Environment.GetEnvironmentVariable ("LANG");
+
+				// This will **not** call `setenv(3)`, but it should be fine since we're after the managed code here, not native
+				Environment.SetEnvironmentVariable ("LANG", langEnvironmentVariable);
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest2.cs:117]: I'm concerned about "thread safety" here; we're modifying global state here, so this test needs to be run "alone" lest we anger the deity of randomness.  Unfortunately, I don't think we're safe, as this fixture is `[Parallelizable (ParallelScope.Children)]`.

Perhaps all we need to do is add `[NonParallelizable]` to this test method?

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest2.cs [hunk 2]
```diff
@@ -105,14 +105,30 @@ public void CompressedWithoutLinker ()
...
+			}
+
+			try {
+				var proj = new XamarinAndroidApplicationProject {
+					IsRelease = isRelease,
+					TargetFrameworkVersion = tfv,
+				};
+				using (var b = CreateApkBuilder ()) {
+					Assert.IsTrue (b.Build (proj), "Build should have succeeded.");
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest2.cs:126]: Can we use the `Build()` parameter here to pass environment variables in?

Example: https://github.com/xamarin/xamarin-android/blob/90afbdd2b89fc0b46ffcd81cb55ad2b85a337bc7/tests/MSBuildDeviceIntegration/Tests/InstallTests.cs#L460

**Review by jonathanpeppers — APPROVED** [review]
LGTM if the new test passes. 👍

**Comment by jonpryor** [comment]
Draft commit message:

```markdown
Fixes: https://github.com/xamarin/xamarin-android/issues/7939

Context: https://github.com/dotnet/runtime/issues/13363
Context: https://www.unicode.org/charts/PDF/U2200.pdf (page 2)
Context: 5271f3e109a2242caa3bd7801dfad9c1683488d7

The LLVM IR generator (5271f3e1) outputs a number of integer values,
relying on the standard `int.ToString()` method to convert the value
to a string.  However, since .NET 6, this conversion is culture-
sensitive and in certain cultures (list below) it will output the
minus sign not as the standard ASCII 0x2D character `-` but rather as
the Unicode character \u2212 `−` (mathematical operator "minus").

This breaks LLVM LLC:

	llc: environment.arm64-v8a.ll:554:7: error: expected value token
	  stderr |                 i32 −1, ; apk_fd

Fix the problem by using invariant culture when converting ***all***
integers and unknown objects to strings.  The reason all of them are
converted in this way is to avoid future changes to ICU and/or .NET
to-string conversion that might affect the resulting integer format.

Workaround: export `$LANG` to a locale that uses 0x2D `-` for
negation, e.g. `LANG=C`.

Locales/cultures affected by this issue are as follows:

  * et
  * et-EE
  * eu
  * eu-ES
  * fi
  * fi-FI
  * fo
  * fo-DK
  * fo-FO
  * gsw
  * gsw-CH
  * gsw-FR
  * gsw-LI
  * hr
  * hr-BA
  * hr-HR
  * ksh
  * ksh-DE
  * lt
  * lt-LT
  * nb
  * nb-NO
  * nb-SJ
  * nn
  * nn-NO
  * no
  * rm
  * rm-CH
  * se
  * se-FI
  * se-NO
  * se-SE
  * sl
  * sl-SI
  * sv
  * sv-AX
  * sv-FI
  * sv-SE

Update `BuildTest2.BuildBasicApplication()` to export
`LANG=sv_SE.UTF-8` as part of the `dotnet build` command to test this
scenario on macOS and Linux.  (This change is ignored on Windows.)
```

**Comment by jonpryor** [comment]
@grendello: will my above suggested workaround of `LANG=C` work?  Or would we need `LANG=en_US.UTF-8`?  (Why suggest `LANG=C` at all?  To be *ever so slightly less* US-centric.  Which is probably laughable, but still.)

**Comment by grendello** [comment]
@jonpryor yes, `LANG=C` should work fine, as it will be as ASCII-based as `en_US.UTF-8` is and should be available on all Unix systems.

**Comment by grendello** [comment]
@jonpryor the commit message needs to be updated with the full list of affected cultures (see the OP here) and it needs to mention that it's not just one Unicode character, but several, that replace the ASCII hyphen as the "minus" sign.

**Pull request merged**